### PR TITLE
feat(virtualization-api): first implementation

### PR DESCRIPTION
**Pull request by yaroslavborbat**
## Description
Added a new component virtualization-api - extension API server.

## Why do we need it, and what problem does it solve?
Using the api server, we can connect to virtual machines and perform various operations. Currently implemented: console, vnc, port forwarding

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

### CODE REVIEW

File: images/virtualization-artifact/pkg/tls/certManager/filesystem/file-cert-manager.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+		case <-f.stopCh:
+			break sync
+		}
+	}
+}
+
+func (f *FileCertificateManager) Stop() {
+	f.certAccessLock.Lock()
+	defer f.certAccessLock.Unlock()
+	if f.stopped {
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/tls/certManager/filesystem/file-cert-manager.go:104]: No need to use f.stopped

	select {
	case <-f.stopCh:
	default:
		close(f.stopCh)
	}

File: images/virtualization-artifact/pkg/apiserver/server/config.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	if c.Apiserver == nil {
+		errs = append(errs, fmt.Errorf(".Apiserver is required. %w", ErrConfigInvalid))
+	}
+	if c.Rest == nil {
+		errs = append(errs, fmt.Errorf(".Rest is required. %w", ErrConfigInvalid))
+	}
+	return errs
+}
+
+func (c Config) Complete() (*server, error) {
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/apiserver/server/config.go:55]: Public method returns private struct

File: images/virtualization-artifact/pkg/apiserver/server/server.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/deckhouse/virtualization-controller/pkg/tls/certManager"
+)
+
+func NewServer(
+	virtualMachines cache.Controller,
+	apiserver *genericapiserver.GenericAPIServer,
+	proxyCertManager certManager.CertificateManager,
+) *server {
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/apiserver/server/server.go:14]: Public returns private

File: images/virtualization-artifact/pkg/apiserver/server/config.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	if err := api.Install(vmInformer.Lister(), genericServer, c.Kubevirt, proxyCertManager); err != nil {
+		return nil, err
+	}
+
+	s := NewServer(
+		vmInformer.Informer(),
+		genericServer,
+		proxyCertManager,
+	)
+	if err != nil {
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/apiserver/server/config.go:79]: Redundant err check

File: images/virtualization-artifact/pkg/apiserver/server/config.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+	if c.ProxyClientKeyFile == "" {
+		errs = append(errs, fmt.Errorf(".ProxyClientKeyFile is required. %w", ErrConfigInvalid))
+	}
+	if c.Apiserver == nil {
+		errs = append(errs, fmt.Errorf(".Apiserver is required. %w", ErrConfigInvalid))
+	}
+	if c.Rest == nil {
+		errs = append(errs, fmt.Errorf(".Rest is required. %w", ErrConfigInvalid))
+	}
+	return errs
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/apiserver/server/config.go:52]: What about errors.Join ?

File: images/virtualization-artifact/pkg/apiserver/registry/vm/storage/util.go [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+)
+
+func nameFor(fs fields.Selector) (string, error) {
+	if fs == nil {
+		fs = fields.Everything()
+	}
+	name := ""
+	if value, found := fs.RequiresExactMatch("metadata.name"); found {
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/apiserver/registry/vm/storage/util.go:15]: name, found := fs.RequiresExactMatch("metadata.name")
if !fs.Empty() {
	return "", fmt.Errorf("field label not supported: %s", fs.Requirements()[0].Field)
}

File: images/virtualization-artifact/pkg/apiserver/registry/vm/storage/storage.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+	if err != nil {
+		return nil, apierrors.NewInternalError(err)
+	}
+
+	filtered := &virtv2.VirtualMachineList{}
+	filtered.Items = make([]virtv2.VirtualMachine, 0, len(items))
+	for _, manifest := range items {
+		var vm *virtv2.VirtualMachine
+		var ok bool
+		if vm, ok = manifest.(*virtv2.VirtualMachine); !ok || vm == nil {
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/apiserver/registry/vm/storage/storage.go:127]: vm, ok := manifest.(*virtv2.VirtualMachine)

File: images/virtualization-artifact/pkg/apiserver/registry/vm/storage/storage.go [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	genericreq "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/client-go/tools/cache"
+
+	virtv2 "github.com/deckhouse/virtualization-controller/api/core/v1alpha2"
+	rest2 "github.com/deckhouse/virtualization-controller/pkg/apiserver/registry/vm/rest"
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/apiserver/registry/vm/storage/storage.go:17]: rename rest2

File: images/virtualization-artifact/pkg/apiserver/registry/vm/rest/stream.go [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+		return nil, err
+	}
+	vm := obj.(*virtv2.VirtualMachine)
+	if vm == nil {
+		return nil, fmt.Errorf("unexpected object type: %#v", vm)
+	}
+	return vm, nil
+}
+
+func streamParams(_ url.Values, opts runtime.Object) error {
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/apiserver/registry/vm/rest/stream.go:94]: remove _ url.Values parameter?

File: images/virtualization-artifact/cmd/virtualization-api/app/start.go [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+	nfs := opts.Flags()
+	for _, f := range nfs.FlagSets {
+		fs.AddFlagSet(f)
+	}
+	local := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+	logs.AddGoFlags(local)
+	nfs.FlagSet("logging").AddGoFlagSet(local)
+
+	usageFmt := "Usage:\n  %s\n"
+	cols, _, _ := term.TerminalSize(cmd.OutOrStdout())
```
Comment (Isteb4k) [inline images/virtualization-artifact/cmd/virtualization-api/app/start.go:39]: Missed err check

File: images/virtualization-artifact/cmd/virtualization-api/app/start.go [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+	cmd.SetUsageFunc(func(cmd *cobra.Command) error {
+		fmt.Fprintf(cmd.OutOrStderr(), usageFmt, cmd.UseLine())
```
Comment (Isteb4k) [inline images/virtualization-artifact/cmd/virtualization-api/app/start.go:41]: Missed error check

File: images/virtualization-artifact/cmd/virtualization-api/app/options/options.go [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+	"k8s.io/client-go/pkg/version"
+	"k8s.io/client-go/rest"
+	"k8s.io/component-base/cli/flag"
+	"k8s.io/component-base/logs"
+	logsapi "k8s.io/component-base/logs/api/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+
+	"github.com/deckhouse/virtualization-controller/pkg/apiserver/api"
+	generatedopenapi "github.com/deckhouse/virtualization-controller/pkg/apiserver/api/generated/openapi"
+	rest2 "github.com/deckhouse/virtualization-controller/pkg/apiserver/registry/vm/rest"
```
Comment (Isteb4k) [inline images/virtualization-artifact/cmd/virtualization-api/app/options/options.go:20]: rest2 to rename

File: images/virtualization-artifact/api/client/kubecli/websocket.go [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+	var msgType int
+	var err error
+	for {
+		if s.reader == nil {
+			msgType, s.reader, err = s.conn.NextReader()
+		} else {
+			msgType = websocket.BinaryMessage
+		}
+		if err != nil {
+			s.reader = nil
```
Comment (Isteb4k) [inline images/virtualization-artifact/api/client/kubecli/websocket.go:99]: It looks like s.reader is already nil

**Review by Isteb4k — CHANGES_REQUESTED** [review]

File: images/virtualization-artifact/pkg/tls/util/util.go [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+		if err != nil {
+			return certs, err
+		}
+
+		certs = append(certs, cert)
+		ok = true
+	}
+
+	if !ok {
+		return certs, errors.New("data does not contain any valid RSA or ECDSA certificates")
```
[the remark this answers is not in the document]
Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/tls/util/util.go:35]: It's ok

File: images/virtualization-artifact/pkg/tls/util/util.go [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+		cert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			return certs, err
+		}
+
+		certs = append(certs, cert)
+		ok = true
+	}
+
+	if !ok {
```
[the remark this answers is not in the document]
Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/tls/util/util.go:34]: it's ok

File: images/virtualization-artifact/pkg/tls/certManager/filesystem/file-cert-manager.go [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+		stopCh:             make(chan struct{}, 1),
+		errorRetryInterval: 1 * time.Minute,
+	}
+}
+
+func (f *FileCertificateManager) Start() {
+	objectUpdated := make(chan struct{}, 1)
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		klog.Errorf("failed to create an inotify watcher: %v", err)
```
[the remark this answers is not in the document]
Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/tls/certManager/filesystem/file-cert-manager.go:40]: no, %w for fmt.Errorf
its logs

File: images/virtualization-artifact/pkg/tls/certManager/filesystem/file-cert-manager.go [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+	}
+	defer watcher.Close()
```
[the remark this answers is not in the document]
Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/tls/certManager/filesystem/file-cert-manager.go:42]: gek. mb we don't want?

File: images/virtualization-artifact/pkg/apiserver/server/config.go (hunk 2 above)
Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/apiserver/server/config.go:55]: Yes. It is assumed that you can create a server only server.Config.

File: images/virtualization-artifact/pkg/apiserver/server/config.go (hunk 4 above)
Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/apiserver/server/config.go:79]: its ok

File: images/virtualization-artifact/pkg/apiserver/registry/vm/storage/storage.go (hunk 7 above)
Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/apiserver/registry/vm/storage/storage.go:127]: its ok

File: images/virtualization-artifact/pkg/apiserver/registry/vm/storage/storage.go (hunk 8 above)
Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/apiserver/registry/vm/storage/storage.go:17]: its ok

File: images/virtualization-artifact/pkg/apiserver/registry/vm/rest/stream.go (hunk 9 above)
Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/apiserver/registry/vm/rest/stream.go:94]: No. Its function should mutate the url.Values.

File: images/virtualization-artifact/Taskfile.dist.yaml [hunk 18]
```diff
@@ -443,6 +444,25 @@ tasks:
...
+        sh: git rev-parse "HEAD^{commit}" 2>/dev/null
+      GIT_TAG:
+        sh: git describe --abbrev=0 --tags 2>/dev/null
+      BUILD_DATE:
+        sh: date -u +'%Y-%m-%dT%H:%M:%SZ'
+    cmds:
+      - |
+        go build -mod=readonly -trimpath \
+        -ldflags "-w -X $PKG/version.gitVersion=$GIT_TAG -X $PKG/version.gitCommit=$GIT_COMMIT -X $PKG/version.buildDate=$BUILD_DATE" \
+        -o "$REPO_DIR/bin/$BINARY_NAME" $REPO_DIR/cmd/virtualization-api
```
[the remark this answers is not in the document]
Comment (yaroslavborbat) [reply images/virtualization-artifact/Taskfile.dist.yaml:464]: no, its equal. remove REPO_DIR

File: images/virtualization-artifact/cmd/virtualization-api/app/start.go (hunk 11 above)
Comment (yaroslavborbat) [reply images/virtualization-artifact/cmd/virtualization-api/app/start.go:41]: it doesn't matter

File: images/virtualization-artifact/pkg/apiserver/server/config.go (hunk 5 above)
Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/apiserver/server/config.go:52]: its ok

File: images/virtualization-artifact/api/client/kubecli/websocket.go (hunk 13 above)
Comment (yaroslavborbat) [reply images/virtualization-artifact/api/client/kubecli/websocket.go:99]: mb is an unnecessary check, but I don't want to change the kubevirt logics in this

File: images/virtualization-artifact/pkg/tls/certManager/filesystem/file-cert-manager.go (hunk 1 above)
Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/tls/certManager/filesystem/file-cert-manager.go:104]: No. if you call Stop more than 1 time you will panic: close of closed channel

File: images/virtualization-artifact/pkg/tls/certManager/filesystem/file-cert-manager.go [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+	}
+
+	go func() {
+		for {
+			select {
+			case _, ok := <-watcher.Events:
+				if !ok {
+					return
+				}
+				select {
```
[the remark this answers is not in the document]
Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/tls/certManager/filesystem/file-cert-manager.go:64]: There's a bit of magic in it. objectupdater is a buffered channel. Therefore, we may not react to file events if we haven't rotate the certificates yet.

File: images/virtualization-artifact/pkg/apiserver/registry/vm/storage/util.go (hunk 6 above)
Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/apiserver/registry/vm/storage/util.go]: its ok

File: images/virtualization-artifact/pkg/tls/certManager/filesystem/file-cert-manager.go (hunk 1 above)
Comment (Isteb4k) [reply images/virtualization-artifact/pkg/tls/certManager/filesystem/file-cert-manager.go:104]: > No. if you call Stop more than 1 time you will panic: close of closed channel

There will be no panic. case <-f.stopCh will be called (with a closing flag). The default case with the channel closure will not be called.

**Review by Isteb4k — APPROVED** [review]

File: hooks/lib/certificate/parse.py [hunk 20]
```diff
@@ -27,13 +27,17 @@ def parse_key(key: str) -> crypto.PKey:
 
 
 def get_certificate_san(crt: crypto.X509) -> list[str]:
-    san = ''
+    san = ""
     ext_count = crt.get_extension_count()
     for i in range(0, ext_count):
         ext = crt.get_extension(i)
-        if 'subjectAltName' in str(ext.get_short_name()):
+        if "subjectAltName"in str(ext.get_short_name()):
```
Comment (diafour) [inline hooks/lib/certificate/parse.py:34]:
```suggestion
        if "subjectAltName" in str(ext.get_short_name()):
```

File: images/virtualization-artifact/api/core/register.go [hunk 21]
```diff
@@ -0,0 +1,4 @@
+package core
+
+// GroupName is the group name use in this package
```
Comment (diafour) [inline images/virtualization-artifact/api/core/register.go:3]:
```suggestion
// GroupName is the group name for resources defined in this package
```

File: images/virtualization-artifact/hack/mirrord.sh [hunk 22]
```diff
@@ -1,33 +1,67 @@
 #!/bin/bash
 set -eo pipefail
 
-function usage { echo "Usage: $0 [run/wipe]" ; exit 1; }
+function usage {
+    cat <<EOF
+  "Usage: $0 [run/wipe] --app=<?> --namespace=<?> --deployment=<?>"
```
Comment (diafour) [inline images/virtualization-artifact/hack/mirrord.sh:6]:
```suggestion
Usage: $0 COMMAND OPTIONS

Commands:
  run    Run locally executed application in a cluster environment.
         Arguments:
         --app           Path to main.go
         --namespace     Namespace of deployment
         --deployment    Deployment where application should be injected
         --flags         Arguments for application.
         
  wipe   Stop and cleanup.
         Arguments:
         --namespace     Namespace of deployment
         --deployment    Deployment where application was injected
```

File: images/virtualization-artifact/pkg/apiserver/server/config.go (hunk 2 above)
Comment (diafour) [reply images/virtualization-artifact/pkg/apiserver/server/config.go:55]: I agree with @Isteb4k. The are no reasons to return private from public constructor. For example, there is no such issue with FileCertificateManager in images/virtualization-artifact/pkg/tls/certManager/filesystem/file-cert-manager.go.

**Pull request merged**